### PR TITLE
 Assert change - replaced assert with if and raised an appropriate exception

### PR DESCRIPTION
**Pull request by ashwini-orchestral**
Replace assert with explicit if check and raised an appropriate exception. The assert statement when run in production (non-debug) code will skip the assert. Although the asserts here are not used in check before running secure code, we want to practice safe coding and only use assert statement in unit tests and not in actual product code.

### CODE REVIEW

File: st2reactor/st2reactor/rules/enforcer.py
Comment (m4dcoder) [inline st2reactor/st2reactor/rules/enforcer.py:164]: Please fix spelling of tracedb

**Review by m4dcoder — CHANGES_REQUESTED** [review]
1. Please check the details of the failed python3.6 job, fix the issue, and get the tests to pass.
2. Please go thru the error messages and make sure they are all consistently capitalized, proper sentence, and ends with a period.
2. Please fix typo.

File: st2reactor/st2reactor/garbage_collector/base.py
Comment (m4dcoder) [inline st2reactor/st2reactor/garbage_collector/base.py:199]: The reverse of `<` should be `>=`.

File: st2reactor/st2reactor/garbage_collector/base.py
Comment (m4dcoder) [inline st2reactor/st2reactor/garbage_collector/base.py:221]: The reverse of `<` should be `>=`.

File: st2reactor/st2reactor/garbage_collector/base.py
Comment (m4dcoder) [inline st2reactor/st2reactor/garbage_collector/base.py:246]: The reverse of `<` should be `>=`.

**Review by m4dcoder — CHANGES_REQUESTED** [review]

**Review by m4dcoder — APPROVED** [review]

File: st2api/st2api/controllers/v1/triggers.py
Comment (blag) [inline st2api/st2api/controllers/v1/triggers.py:329]:
```suggestion
                                    f' (was {type(self.payload)}).')
```

File: st2common/st2common/bootstrap/actionsregistrar.py
Comment (blag) [inline st2common/st2common/bootstrap/actionsregistrar.py:214]: What is the type of `pack_base_paths`?

File: st2common/st2common/bootstrap/aliasesregistrar.py
Comment (blag) [inline st2common/st2common/bootstrap/aliasesregistrar.py:195]: Same here - help the user debug what happened by including what was the type of `packs_base_path` in the exception message.

File: st2common/st2common/bootstrap/configsregistrar.py
Comment (blag) [inline st2common/st2common/bootstrap/configsregistrar.py:153]: Here too

File: st2common/st2common/models/api/keyvalue.py
Comment (blag) [inline st2common/st2common/models/api/keyvalue.py:192]: What are those values?

File: st2common/st2common/runners/base.py
Comment (blag) [inline st2common/st2common/runners/base.py:293]: I don't think this will do what you want it to do.

```suggestion
                raise ValueError('The entry point value "%s" '
                                 'does not match with git worktree path "%s".' %
```

File: st2common/st2common/runners/base.py
Comment (blag) [inline st2common/st2common/runners/base.py:383]:
```suggestion
            raise ValueError('The worktree path "%s" is not within /tmp.' %
```

File: st2common/st2common/runners/base.py
Comment (blag) [inline st2common/st2common/runners/base.py:387]:
```suggestion
            raise ValueError('The worktree path is not within /tmp/"%s".' %
                             (self.WORKTREE_DIRECTORY_PREFIX))
```

File: st2common/st2common/services/trace.py
Comment (blag) [inline st2common/st2common/services/trace.py:58]: This exception message should be improved to indicate that `trace_context` can be either a `TraceContext` object or a dictionary.

File: st2common/st2common/services/triggers.py
Comment (blag) [inline st2common/st2common/services/triggers.py:257]:
```suggestion
        raise ValueError(f'The trigger has a value that is not a dictionary (was {type(trigger)}).')
```

File: st2common/st2common/services/triggers.py
Comment (blag) [inline st2common/st2common/services/triggers.py:411]:
```suggestion
        raise ValueError(f'The trigger has a value that is not a dictionary (was {type(trigger_type)}).')
```

File: st2common/st2common/transport/reactor.py
Comment (blag) [inline st2common/st2common/transport/reactor.py:96]:
```suggestion
        if payload and not isinstance(payload, dict):
```

File: st2common/st2common/transport/reactor.py
Comment (blag) [inline st2common/st2common/transport/reactor.py:98]:
```suggestion
        if trace_context and not isinstance(trace_context, TraceContext):
```

File: st2common/st2common/util/sandboxing.py
Comment (blag) [inline st2common/st2common/util/sandboxing.py:115]:
```suggestion
            raise ValueError(f'The prefix "{sys_prefix}" is not found in "{site_packages_dir}".')
```

**Review by blag — CHANGES_REQUESTED** [review]
Better, but still some changes to make.

There are additional places that I did not mark that need to include the type of the object under inspection in the exception message. Please improve all exception messages that are like that.

File: st2common/st2common/transport/reactor.py
Comment (blag) [inline st2common/st2common/transport/reactor.py:97]: What is the type of `payload`?

Same question for `trace_context` below.

File: st2common/st2common/util/crypto.py
Comment (blag) [inline st2common/st2common/util/crypto.py:83]:
```suggestion
if DEFAULT_AES_KEY_SIZE < MINIMUM_AES_KEY_SIZE:
```

File: st2common/st2common/util/crypto.py
Comment (blag) [inline st2common/st2common/util/crypto.py:212]: What is the type of `encrypt_key`?

File: st2common/st2common/util/crypto.py
Comment (blag) [inline st2common/st2common/util/crypto.py:214]: What is the type of `plaintext`?

File: st2common/st2common/util/crypto.py
Comment (blag) [inline st2common/st2common/util/crypto.py:272]: What is the type of `decrypt_key`?

File: st2common/st2common/util/crypto.py
Comment (blag) [inline st2common/st2common/util/crypto.py:274]: What is the type of `ciphertext`?

File: st2common/st2common/util/templating.py
Comment (blag) [inline st2common/st2common/util/templating.py:44]: What is the type of `value`?

File: st2reactor/st2reactor/container/sensor_wrapper.py
Comment (blag) [inline st2reactor/st2reactor/container/sensor_wrapper.py:365]: What is the type of `parent_args`?

File: st2reactor/st2reactor/garbage_collector/base.py
Comment (blag) [inline st2reactor/st2reactor/garbage_collector/base.py:201]: Please do not combine f-strings with percent interpolation.

Same issue with the exception messages further down on lines 222/223 and 247/248.

**Review by blag — CHANGES_REQUESTED** [review]
Still need to improve exception messages to make them more user friendly.

**Comment by cognifloyd** [comment]
We just reformatted the code with `black`. (Hooray!) And this PR got caught in the cross fire. (Arrgh!)
You'll need to merge master into your branch, resolve the conflicts (ouch - there'll be a lot of them for this PR - sorry!), and reformat with black (`pre-commit install` is a good idea).

PS: OOH! f-strings... excellent!

File: st2common/st2common/content/utils.py
Comment (blag) [inline st2common/st2common/content/utils.py:301]: Subtle typo - just needs an extra space.

```suggestion
            f" to prevent directory traversal {result}."
```

File: st2common/st2common/models/api/keyvalue.py
Comment (blag) [inline st2common/st2common/models/api/keyvalue.py:188]: Same subtle bug.

```suggestion
                    f" same original encrypted value {original_value}."
```

File: st2common/st2common/runners/base.py
Comment (blag) [inline st2common/st2common/runners/base.py:298]:
```suggestion
                    f"The entry point value {self.entry_point}"
                    " does not start with the git worktree path"
                    f" {self.git_worktree_path}."
```

File: st2common/st2common/runners/base.py
Comment (blag) [inline st2common/st2common/runners/base.py:398]:
```suggestion
                f"The worktree path is not within /tmp (was {type(worktree_path)})."
```

File: st2common/st2common/transport/announcement.py
Comment (blag) [inline st2common/st2common/transport/announcement.py:72]:
```suggestion
                f"The trace context has a value that is not a"
                " NoneType or dict or TraceContext"
                f" (was {type(trace_context)})."
```

File: st2reactor/st2reactor/garbage_collector/base.py
Comment (blag) [inline st2reactor/st2reactor/garbage_collector/base.py:237]:
```suggestion
                f" later than now in UTC ({utc_now})."
```

File: st2reactor/st2reactor/garbage_collector/base.py
Comment (blag) [inline st2reactor/st2reactor/garbage_collector/base.py:269]:
```suggestion
                f" later than now in UTC ({utc_now})."
```

File: st2reactor/st2reactor/garbage_collector/base.py
Comment (blag) [inline st2reactor/st2reactor/garbage_collector/base.py:300]:
```suggestion
                f" later than now in UTC ({utc_now})."
```

**Review by blag — APPROVED** [review]
Approved, after suggestions are merged in.

The f-strings that don't contain any formatting parameters is a little weird, but whatever.

File: contrib/runners/action_chain_runner/action_chain_runner/action_chain_runner.py
Comment (m4dcoder) [reply contrib/runners/action_chain_runner/action_chain_runner/action_chain_runner.py:963]: This is introduced by black formatting. It's fixe.

File: contrib/runners/python_runner/python_runner/python_action_wrapper.py
Comment (m4dcoder) [reply contrib/runners/python_runner/python_runner/python_action_wrapper.py:368]: This is introduced by black format. Black auto concat two lines of code. Fixed.

File: contrib/runners/python_runner/python_runner/python_action_wrapper.py
Comment (m4dcoder) [reply contrib/runners/python_runner/python_runner/python_action_wrapper.py:327]: Same explanation. Fixed.

File: st2api/st2api/controllers/v1/actionexecutions.py
Comment (m4dcoder) [reply st2api/st2api/controllers/v1/actionexecutions.py:451]: Fixed.

File: st2common/st2common/runners/base.py
Comment (m4dcoder) [reply st2common/st2common/runners/base.py:398]: Fixed

File: st2common/st2common/transport/announcement.py
Comment (m4dcoder) [reply st2common/st2common/transport/announcement.py:72]: Fixed

File: st2reactor/st2reactor/garbage_collector/base.py
Comment (m4dcoder) [reply st2reactor/st2reactor/garbage_collector/base.py:237]: Fixed

File: st2reactor/st2reactor/garbage_collector/base.py
Comment (m4dcoder) [reply st2reactor/st2reactor/garbage_collector/base.py:269]: Fixed

File: st2reactor/st2reactor/garbage_collector/base.py
Comment (m4dcoder) [reply st2reactor/st2reactor/garbage_collector/base.py:300]: Fixed

File: st2common/st2common/runners/base.py
Comment (m4dcoder) [reply st2common/st2common/runners/base.py:298]: Fixed

**Comment by Kami** [comment]
Just to add quick context here from my side - I'm fine with this change and as pointed out in the PR description, asserts were only used to guard against programmer errors and not end users errors or invalid data (for those things, we use exceptions) - so being stripped out for a production build is not an issue.

Having said that, using exceptions everywhere is fine as well (and indeed perhaps a bit safer is we would miss ``assert`` being used incorrectly in a PR review).

**Pull request merged**
